### DISCUSSION TITLE
Problems with AOT.

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#16837] [Q&A]
I have an Avalonia program that is running well and I want to switch it to AOT mode.

I don't know much about AOT, and this is my first time trying to publish an Avalonia program for AOT.

I encountered a problem. I successfully published the program in AOT mode, but after opening the program, the program UI was blank.

I checked the information on the Internet and found that this might be the reason:
I wrote an independent Avalonia UI library and stored it in an independent dll file. My Avalonia program will reference this dll file. According to what is said on the Internet, my dll file must also implement AOT.

So now the question is, after this dll file implements AOT, how can I reference the TemplatedControl, UserControl, ResourceDictionary, Style, Converter in this dll file in my Avalonia program?

**Comment by thevortexcloud** [comment]
AOT is generally not something you just switch on. It requires you to think pretty hard about how your app is going to work in a lot of cases, as a lot of things just wont work with it or require workarounds.

That said, once you have AOT setup for the most part you can just use it like a regular .NET project. The compiler will deal with the rest.

So my question is: Where are you compiling your app and where are you trying to run it? AOT converts your app into a _native_ app. The result is basically the same as writing your app in something like C. 

Also how are you trying to display your controls? If you are trying to load an assembly using reflection you probably will run into problems.

**Reply by kyuranger** [reply]
The reason why I want to convert the program to AOT compilation is not a performance issue.

My program runs on both Windows and Linux DRM across platforms. Since I use ordinary compilation, some criminals can easily decompile my software, obtain the database password and tamper with the data. Of course, the safest way to protect the database is through the server, but since my product is for stand-alone use, it is not possible to use server transmission.

Moreover, even if this problem can be solved through legal means in theory, it is very difficult to obtain evidence.

I know that a simple way to solve this problem is to use an obfuscator. However, not all obfuscators on the market support .NET 8 on the one hand, and basically do not support Linux DRM on the other hand. Moreover, I have tried many trial versions of obfuscators, and it seems that most of them have problems even compiling after obfuscation on the Windows platform. I am not sure whether the obfuscator needs to support Avalonia, and there are almost no tutorials or articles on this aspect.

So, I thought of AOT compilation to solve this problem. Although I understand that even if AOT compilation is adopted, it can still be cracked by some hackers, but at least the difficulty of cracking is much higher.

I have seen many people mention in other forums that they are using Avalonia for AOT compilation. I have no experience with AOT compilation, but from their chats I thought AOT compilation would be very simple. However, as mentioned above, I encountered many problems, but people in other forums could not give any solutions, so I raised this question.

**Reply by kyuranger** [reply]
WTF, I just saw today that Microsoft released the latest version of .NET Community Toolkit yesterday(https://devblogs.microsoft.com/dotnet/announcing-the-dotnet-community-toolkit-830/). In the changelog of the latest version, it mentioned the new support for AOT. I have been using .NET Community Toolkit, but I always thought it natively supports AOT.

**Reply by djeikyb** [reply]
fwiw i used CommunityToolkit.Mvvm 8.2.2 with a simple aot app just fine. wouldn't say i exhausted its feature set, but did use:

- ObservablePropertyAttribute
- ObservableObject
- RelayCommandAttribute

**Reply by maxkatz6** [reply]
>  I am not sure whether the obfuscator needs to support Avalonia, and there are almost no tutorials or articles on this aspect.

No, but there is a high chance you will get the same issues with third parties.

> In the changelog of the latest version, it mentioned the new support for AOT

CommunityToolkit.Mvvm always supported AOT.
That's WinUI controls from the same toolkit that are now compatible.

**Reply by kyuranger** [reply]
> > I am not sure whether the obfuscator needs to support Avalonia, and there are almost no tutorials or articles on this aspect.
> 
> No, but there is a high chance you will get the same issues with third parties.
> 
> > In the changelog of the latest version, it mentioned the new support for AOT
> 
> CommunityToolkit.Mvvm always supported AOT. That's WinUI controls from the same toolkit that are now compatible.

Thanks for your answer

**Reply by thevortexcloud** [reply]
> Since I use ordinary compilation, some criminals can easily decompile my software, obtain the database password and tamper with the data

If this is your goal, then you are thinking about things wrong. AOT also wont help with this as your connection information still has to be embedded _somewhere_ in your code for it to work.

The proper solution would be to:

1. Don't embed database credentials in the first place. Generally it is considered to be very bad practice to give people direct access to a database without a very good reason
2. Obtain credentials from somewhere else. Some examples:
    -  Ask the user to provide their own credentials
    - Use some sort of single sign on (EG Windows auth)
    - Remove direct database access entirely and instead use some sort of web API (EG REST) to talk to your database on behalf of your app. This also has the advantage of allowing some sort of user account control (EG restricting what people can access) if you set it up with that in mind.

> I know that a simple way to solve this problem is to use an obfuscator

This is not guaranteed to solve the problem as strings are generally not obfuscated. They _can_ be obfuscated at the cost of extra work at runtime to reassemble them.

Furthermore all an obfuscator does is slow people down. If someone _really_ wants to reverse engineer your app, it wont stop them. There are many tools that exist for the specific purpose of making things like this easier. Also an obsufcator only protects your app at rest. There are possibly many other ways to obtain the same information that would bypass the obfuscation entirely, such as:

* Dumping the app's data that is currently in RAM while it is running 
* Intercepting network traffic
 
Trying to protect yourself against people who have physical access to the machine is a very hard problem to solve, and in most cases is not worth it. In extreme cases, [people will even use a drill to bypass your security.](https://kotaku.com/one-of-the-wildest-console-hacks-ever-1847455427)

**Reply by kyuranger** [reply]
I understand that this is just like a single-player game. No software can be completely crack-proof. Even if single-player games are now changed to online authentication, they can still be cracked. But I just hope to increase the difficulty of cracking as much as possible to avoid the chance of loss.

**Reply by thevortexcloud** [reply]
I understand that, but your approach is not going to work with your goals. As I mentioned, there are alternative ways to achieve the same result without the same downsides. Embedding your database credentials inside your app is the same as directly giving them to the user. I would strongly suggest rethinking your design.

That said, some database libraries are not AOT compatible. Raw ADO typically is though.

**Reply by kyuranger** [reply]
Thanks for your suggestion, to achieve this, I have migrated my ORM from EFCORE to SQLSUGAR (SQLSUGAR supports AOT).

**Comment by kyuranger** [comment]
Tonight I suddenly realized that I had fallen into a serious logical error. In my code, I wrote all the registration code-related codes, TemplatedControl, UserControl, ResourceDictionary, Style, Converter, etc. in one Library. But in fact, I only need to encrypt the registration code-related codes through AOT. Put aside the debate on whether it is feasible to encrypt the code through AOT. In fact, the part I need to AOT is only the registration code-related code. So what I should do is to separate the registration code-related code into an independent library and find a way to AOT it, instead of AOTing all the parts such as TemplatedControl, UserControl, ResourceDictionary, Style, Converter, etc. (this will waste a lot of development time).

**Reply by thevortexcloud** [reply]
I could be wrong, but I am pretty sure you can't directly reference an AOT assembly from a non AOT assembly, as at that point it is no longer a .NET assembly. You can still reference it via a `DllImport` though, because as far as the runtime is concerned, it's a "C" like assembly.

https://stackoverflow.com/questions/75419409/how-to-use-a-dll-generated-in-c-sharp-and-compiled-using-native-aot-in-other-ne

As a result you would still need to AOT compile everything or do hacks using a DllImport.

>  But in fact, I only need to encrypt the registration code-related codes through AOT

As I said before, AOT does not involve any encryption/obfuscation. Someone can still decompile it just as easily as they can a .NET assembly. You could use real encryption though, but you would still need to have a decryption key somewhere.

**Reply by kyuranger** [reply]
BTW, I just tried to AOT my Avalonia program (I found a bug that prevented it from running after AOT). But after fixing the bug and AOTing it, Windows prompted "The system detected an overrun of a stack-based buffer in this application. This overrun could potentially allow a malicious user to gain control of this application."

I checked the information online and it seemed very complicated to solve this problem. So I had to give up the idea of ​​AOTing the entire program again.

**Comment by josephmoresena** [comment]
I think the approach you are looking for is the wrong case for AOT. Some time ago I remember a friend and I were looking for a way to use a native DLL for license validation and generation for an application using certain information from the device and the user input, that was wrong any way.

About native AOT, at some point using NativeAOT for a project in the company I work for I had problems with embedding resources (Themes). The solution in the end was to use rd.xml, try for the moment to keep all the assemblies:

`<Assembly Name="NombreDllAvalonia" Dynamic="Required All"/>`

Which version of Avalonia are you using? If 11.x, include a package reference to Avalonia.Themes.*

**Reply by kyuranger** [reply]
Now I changed my license validation way. I used a third-party dongle, and then encapsulated the dongle-related code into a library and performed AOT. At the same time, some information such as the database password is also dynamic (different for each device), and this information is also controlled by this AOT library.

I am not sure whether my idea is feasible. I have not had time to test it recently.

**Reply by kyuranger** [reply]
The version of Avalonia I am using is the latest version 11.1.3

**Reply by josephmoresena** [reply]
> The version of Avalonia I am using is the latest version 11.1.3

What theme are you using? Can you include the theme assemblies in the rd.xml file? Once you do that, do the elements appear in the Window?
Would it be possible for you to create a repository with an application that simply initializes the window and contains the user controls used?

**Reply by kyuranger** [reply]
I didn't use Avalonia's native theme system, I used a theme system I made myself. All my controls were remade based on templated controls. Then the controls load different colors (static resources) as needed when the program is opened.

**Reply by josephmoresena** [reply]
> I didn't use Avalonia's native theme system, I used a theme system I made myself. All my controls were remade based on templated controls. Then the controls load different colors (static resources) as needed when the program is opened.

You need to preserve all resources/themes even if the program doesn't need them at load time. Can you preserve the entire resource assembly via rd.xml? Can you try? Are you using avares for embedded resources?

**Reply by kyuranger** [reply]
Yes, I am using avares for embedded resources. I will try this later.
